File: src/main/java/me/arasple/mc/trmenu/TrMenuLoader.java
package me.arasple.mc.trmenu;

import io.izzel.taboolib.module.config.TConfig;
import io.izzel.taboolib.module.locale.TLocale;
import io.izzel.taboolib.util.Files;
import me.arasple.mc.trmenu.api.TrMenuAPI;
import me.arasple.mc.trmenu.menu.MenuLoader;
import me.arasple.mc.trmenu.updater.Updater;
import me.arasple.mc.trmenu.utils.Bungees;
import me.arasple.mc.trmenu.utils.FileWatcher;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.HumanEntity;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.logging.Level;

/**
 * @author Arasple
 * @date 2019/12/8 9:09
 */
public class TrMenuLoader {

    void init() {
        if (!TrMenu.SETTINGS.getBoolean("OPTIONS.SILENT", false)) {
            Arrays.asList(
                    "",
                    "§3  ____________________   _____  __________________   ____ ___",
                    "§3\\__    ___\\______   \\ /     \\ \\_   _____/\\      \\ |    |   \\",
                    "§3  |    |   |       _//  \\ /  \\ |    __)_ /   |   \\|    |   /",
                    "§3  |    |   |    |   /    Y    \\|        /    |    |    |  /",
                    "§3  |____|   |____|_  \\____|__  /_______  \\____|__  |______/",
                    "§3                  \\/        \\/        \\/        \\/"
            )
                    .forEach(l -> Bukkit.getConsoleSender().sendMessage(l));
        }
        TLocale.sendToConsole("PLUGIN.LOADING");

        if (installDepend()) {
            return;
        }

        updateConfig();
        Updater.init(TrMenu.getPlugin());
        Bungees.init();
        MenuLoader.init();
        MenuLoader.loadMenus(Bukkit.getConsoleSender());
    }


    void active() {
        TLocale.sendToConsole("PLUGIN.ENABLED", TrMenu.getPlugin().getDescription().getVersion());
    }

    void cancel() {
        Bukkit.getOnlinePlayers().stream().filter(TrMenuAPI::isViewingMenu).forEach(HumanEntity::closeInventory);
        FileWatcher.getWatcher().unregisterAll();
        if (Updater.isAutoUpdate() && Updater.isOld()) {
            String url = "https://arasple.oss-cn-beijing.aliyuncs.com/files/TrMenu.jar";
            Files.downloadFile(url, TrMenu.getPluginFile());
            TrMenu.getPlugin().getLogger().log(Level.INFO, "Successfully downloaded a new version of TrMenu...");
        }
    }

    private boolean installDepend() {
        Plugin plugin = Bukkit.getPluginManager().getPlugin("PlaceholderAPI");
        File jarFile = new File("plugins/PlaceholderAPI.jar");
        String url = "https://api.spiget.org/v2/resources/6245/download";

        if (plugin == null) {
            jarFile.delete();
            TLocale.sendToConsole("PLUGIN.DEPEND.DOWNLOAD", "PlaceholderAPI");
            if (Files.downloadFile(url, jarFile)) {
                TLocale.sendToConsole("PLUGIN.DEPEND.INSTALL", "PlaceholderAPI");
            } else {
                TLocale.sendToConsole("PLUGIN.DEPEND.INSTALL-FAILED", "PlaceholderAPI");
            }
            Bukkit.shutdown();
            return true;
        }
        return false;
    }

    private void updateConfig() {
        TConfig oldCfg = TrMenu.SETTINGS;
        YamlConfiguration newCfg = YamlConfiguration.loadConfiguration(new InputStreamReader(TrMenu.getPlugin().getResource("settings.yml")));
        if (newCfg.getInt("CONFIG-VERSION") == TrMenu.SETTINGS.getInt("CONFIG-VERSION")) {
            return;
        }
        oldCfg.set("CONFIG-VERSION", newCfg.getInt("CONFIG-VERSION"));
        oldCfg.getKeys(true).forEach(key -> {
            if (!newCfg.contains(key)) {
                oldCfg.set(key, null);
            }
        });
        newCfg.getKeys(true).forEach(key -> {
            if (!oldCfg.contains(key)) {
                oldCfg.set(key, newCfg.get(key));
            }
        });
        oldCfg.saveToFile();
    }

}

File: src/main/kotlin/me/arasple/mc/trmenu/TrMenuLoader.kt
package me.arasple.mc.trmenu

import io.izzel.taboolib.TabooLib
import io.izzel.taboolib.Version
import io.izzel.taboolib.module.locale.TLocale
import io.izzel.taboolib.module.locale.TLocaleLoader
import me.arasple.mc.trmenu.modules.command.registerable.RegisterCommands
import me.arasple.mc.trmenu.modules.conf.MenuLoader
import me.arasple.mc.trmenu.modules.function.Shortcuts
import me.arasple.mc.trmenu.modules.function.hook.*
import me.arasple.mc.trmenu.util.Watchers
import me.clip.placeholderapi.PlaceholderAPIPlugin
import org.bukkit.Bukkit

/**
 * @author Arasple
 * @date 2020/2/29 17:27
 */
class TrMenuLoader {

    fun init() {
        TrMenu.SETTINGS.listener { register() }
        if (!TrMenu.SETTINGS.getBoolean("Options.Hide-Logo", false)) {
            printLogo()
            TLocale.sendToConsole("PLUGIN.LOADING", Version.getBukkitVersion())
        }
        register()
    }

    fun active() {
        hook()
        TLocale.sendToConsole("PLUGIN.LOADED", TrMenu.plugin.description.version)
        MenuLoader.loadMenus()
    }

    fun cancel() {
        PlaceholderAPIPlugin.getInstance().localExpansionManager.findExpansionByIdentifier("trmenu").ifPresent {
            it.unregister()
        }
        Watchers.watcher.unregisterAll()
    }

    private fun hook() {
        HookCMI.init()
        HookCronus.init()
        HookHeadDatabase.init()
        HookPlayerPoints.init()
        HookSkinsRestorer.init()
    }

    private fun register() {
        syncLocale()
        RegisterCommands.load()
        Shortcuts.load()
    }

    private fun syncLocale() {
        if (TLocaleLoader.getLocalPriorityFirst(TrMenu.plugin) != "zh_CN" && TLocaleLoader.getLocalPriorityFirst(TabooLib.getPlugin()) == "zh_CN") {
            TabooLib.getConfig().also {
                it.set("LOCALE.PRIORITY", listOf("en_US", "zh_CN"))
                it.saveToFile()
            }
        }
    }

    private fun printLogo() = arrayOf(
        "§8  ___________         _____                        ________",
        "§8    \\__    ___/______  /     \\   ____   ____  __ __  \\_____  \\",
        "§8    |    |  \\_  __ \\/  \\ /  \\_/ __ \\ /    \\|  |  \\  /  ____/",
        "§8    |    |   |  | \\/    Y    \\  ___/|   |  \\  |  / /       \\",
        "§8    |____|   |__|  \\____|__  /\\___  >___|  /____/  \\_______ \\",
        "§8    \\/     \\/     \\/                \\/",
        "                                                      "
    ).let {
        it.forEachIndexed { index, raw ->
            if (raw.isNotBlank()) {
                val line = raw.toCharArray()
                val width = (3..6).random()
                var randomIndex: Int
                do {
                    randomIndex = (2..line.size - width).random()
                } while (String(line.copyOfRange(randomIndex, randomIndex + width)).isBlank())
                val replace = String(line.copyOfRange(randomIndex, randomIndex + width))
                it[index] = String(line).replaceFirst(replace, "§${arrayOf('9', 'b', '3').random()}$replace§8")
            }
        }
        Bukkit.getConsoleSender().sendMessage(it)
    }


}